### PR TITLE
feat(ecs-patterns): Add Ephemeral Storage Support For Ecs Patterns

### LINKED ISSUE #18105: (aws-ecs-patterns): add support for Fargate ephemeral storage
### Description

In April 2021, Fargate added support for [requesting additional ephemeral storage](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/fargate-task-storage.html#fargate-task-storage-pv14).

Support for this feature has [been added](https://github.com/aws/aws-cdk/commit/f1bf935c47006096b33fb7bf0c847ffab9230870#diff-dcfbc499b4d3c10afcd4e63ad0e4ecc54df2464e45af67f1fdae69d3fa2d43a0) to the `aws-ecs` package, but is still not supported by the `aws-ecs-patterns` package.

### Use Case

`aws-ecs-patterns` supports the following patterns:

* ApplicationLoadBalancedFargateService
* ApplicationMultipleTargetGroupsFargateService
* NetworkLoadBalancedFargateService
* NetworkMultipleTargateGroupsFargateService
* QueueProcessingFargateService
* ScheduledFargateTask

Users of these patterns are currently unable to request additional ephemeral storage beyond the 20 GiB provided by default. By adding support for this new feature, users will be able to request up to 200 GiB of ephemeral storage for their Fargate services.

### Proposed Solution

The proposal is to follow the precedent set by the `aws-ecs` package:

```ts
const fargateTaskDefinition = new ecs.FargateTaskDefinition(this, 'TaskDef', {
  memoryLimitMiB: 512,
  cpu: 256,
  ephemeralStorageGiB: 100
});
```

All the patterns in `aws-ecs-patterns` use under the hood `aws-ecs`'s `FargateTaskDefinition`. Therefore, the proposed implementation will use the same naming convention, and forward the value to `ecs.FargateTaskDefinition`. All validation and setting of default values will be left to `aws-ecs`.

This is what a `FargateScheduledTask` would look like:

```ts
declare const cluster: ecs.Cluster;
const scheduledFargateTask = new ecsPatterns.ScheduledFargateTask(this, 'ScheduledFargateTask', {
  cluster,
  scheduledFargateTaskImageOptions: {
    image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
    memoryLimitMiB: 512,
    ephemeralStorageGiB: 100,
  },
  schedule: appscaling.Schedule.expression('rate(1 minute)'),
  platformVersion: ecs.FargatePlatformVersion.LATEST,
});
```

### Other information

`aws-ecs-patterns` is inconsistent in how it exposes some configuration options. `FargateScheduledTask` puts the cpu/memory options inside `FargateTaskImageOptions`:

```ts
declare const cluster: ecs.Cluster;
const scheduledFargateTask = new ecsPatterns.ScheduledFargateTask(this, 'ScheduledFargateTask', {
  cluster,
  scheduledFargateTaskImageOptions: {
    image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
    cpu: 512,
    memoryLimitMiB: 1024,
  },
  schedule: appscaling.Schedule.expression('rate(1 minute)'),
  platformVersion: ecs.FargatePlatformVersion.LATEST,
});
```

By contrast, all the other patterns (ALB, NLB, Queue) set these values outside of the task image options, as top-level properties:

```ts
declare const cluster: ecs.Cluster;
const loadBalancedFargateService = new ecsPatterns.ApplicationLoadBalancedFargateService(this, 'Service', {
  cluster,
  memoryLimitMiB: 1024,
  cpu: 512,
  taskImageOptions: {
    image: ecs.ContainerImage.fromRegistry("amazon/amazon-ecs-sample"),
  },
});
```

For backwards-compatibility, this proposal will not attempt to address this inconsistency between the patterns, and will follow the existing convention set by each pattern.

### Acknowledge

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by homakk**
Fargate supports a new ephemeral storage option, to request from 20GiB to 200
GiB of task storage. Support for it was added to aws-ecs but not to
aws-ecs-patterns.

This change adds support for ephemeral storage, so the following
patterns can take advantage of it:

- Application load balanced fargate service
- Application multiple target groups fargate service
- Network load balanced fargate service
- Network multiple target groups fargate service
- Queue processing fargate service
- Scheduled fargate task

> Remember to follow the [CONTRIBUTING GUIDE] and [DESIGN GUIDELINES] for any
> code you submit.
>
> [CONTRIBUTING GUIDE]: https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md
> [DESIGN GUIDELINES]: https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md

Closes #18105.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by bvtujo — COMMENTED** [review]
LGTM, shy of a few nits and a test case request

File: packages/@aws-cdk-testing/framework-integ/test/aws-servicecatalog/test/product-stack-snapshots/integservicecatalogproductProductStackHistory4D520FF1.integservicecatalogproductSNSTopicProduct3B51CF591.v1.product.template.json
[the remark this answers is not in the document]
Comment (homakk) [reply packages/@aws-cdk-testing/framework-integ/test/aws-servicecatalog/test/product-stack-snapshots/integservicecatalogproductProductStackHistory4D520FF1.integservicecatalogproductSNSTopicProduct3B51CF591.v1.product.template.json:1]: When I updated integ test and run yarn build this file appeared. I don't know why it came actually.

File: packages/aws-cdk-lib/aws-ecs-patterns/test/fargate/load-balanced-fargate-service.test.ts
[the remark this answers is not in the document]
Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/test/fargate/load-balanced-fargate-service.test.ts:1]: No, In `aws-ecs-patterns` we dont have saparate test files for all the patterns. 
We have only one `scheduled-ecs-tast.test.ts` and `queue-processing-ecs-service.test.ts` which are related to this task in the ecs realted tests folder.

We added that validation (`ephemeralStorageGiB is only supported for Fargate services `) in all the pattern except  queue-processing-ecs-service
Because in queue processing pattern the taskImageOption was not present. We add ephemeral Storage through props.` ephemeralStorageGiB: props.ephemeralStorageGiB,`

 That's why I have added the those test case scenario in only one scheduled-ecs-tast.test.ts file.
 
image

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/ecs/application-load-balanced-ecs-service.ts [hunk 1]
```diff
@@ -108,6 +108,9 @@ export class ApplicationLoadBalancedEc2Service extends ApplicationLoadBalancedSe
       this.taskDefinition = props.taskDefinition;
     } else if (props.taskImageOptions) {
       const taskImageOptions = props.taskImageOptions;
+      if (taskImageOptions.ephemeralStorageGiB) {
+        throw new Error('ephemeralStorageGiB is only supported for Fargate service.');
```
[the remark this answers is not in the document]
Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/ecs/application-load-balanced-ecs-service.ts]: Yes I changed service -> services  in Error message and also updated all other files where this validation added. You can see it in the latest commit.

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts [hunk 2]
```diff
@@ -391,6 +391,15 @@ export interface ApplicationLoadBalancedTaskImageOptions {
   * @default none
   */
   readonly command?: string[];
+
+  /**
+   * The amount (in GiB) of ephemeral storage to be allocated to the task. The maximum supported value is 200 GiB.
+   *
+   * NOTE: This parameter is only supported for tasks hosted on AWS Fargate using platform version 1.4.0 or later.
+   *
+   * @default - Undefined, in which case, the task will receive 20GiB ephemeral storage.
```
Comment (iamhopaul123) [inline packages/aws-cdk-lib/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts:400]:
```suggestion
   * @default - Undefined, in which case, the task will receive 20 GiB ephemeral storage.
```
smae for the other files

**Review by iamhopaul123 — COMMENTED** [review]
LGTM overall just a very nit comment

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/base/scheduled-task-base.ts [hunk 3]
```diff
@@ -112,7 +112,7 @@ export interface ScheduledTaskImageProps {
   readonly logDriver?: LogDriver;
 
   /**
-   * The amount (in GiB) of ephemeral storage to be allocated to the task. The maximum supported value is 200 GiB.
+   * The amount (in GiB) of ephemeral storage to be allocated to the task. The maximum supported value is200 GiB.
```
Comment (iamhopaul123) [inline packages/aws-cdk-lib/aws-ecs-patterns/lib/base/scheduled-task-base.ts:115]: nit: typo

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts (hunk 2 above)
Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/base/application-load-balanced-service-base.ts:400]: I mentioned this as per aws-ecs package. There is no space between 20 and GiB. I
But I made changes for 200 GiB -> 200GiB which I mentioned in the same description, to make it consistent as @iamhopaul123 suggested me in the slack.

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/base/scheduled-task-base.ts (hunk 3 above)
Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/base/scheduled-task-base.ts:115]: Updated already. Which can find in the latest commit

**Review by iamhopaul123 — APPROVED** [review]

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/ecs/application-load-balanced-ecs-service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (bvtujo) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/ecs/application-load-balanced-ecs-service.ts]: I don't understand your explanation why we can't test for this condition. We can add a new test in [test/ec2/l3s-v2.test.ts](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-ecs-patterns/test/ec2/l3s-v2.test.ts) to make sure that this condition is triggered when taskImageOptions with ephemeral storage is specified for an ALB EC2 Service.

File: packages/@aws-cdk-testing/framework-integ/test/aws-servicecatalog/test/product-stack-snapshots/integservicecatalogproductProductStackHistory4D520FF1.integservicecatalogproductSNSTopicProduct3B51CF591.v1.product.template.json
[the remark this answers is not in the document]
Comment (bvtujo) [reply packages/@aws-cdk-testing/framework-integ/test/aws-servicecatalog/test/product-stack-snapshots/integservicecatalogproductProductStackHistory4D520FF1.integservicecatalogproductSNSTopicProduct3B51CF591.v1.product.template.json:1]: Can you remove it from the PR?

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/ecs/application-load-balanced-ecs-service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/ecs/application-load-balanced-ecs-service.ts]: Added testcase for ApplicationLoadBalancedEC2Service and when I gone through those files I thought I can add test cases for NetworkMultipleTargetGroupsEc2Service, ApplicationMultipleTargetGroupsEc2Service as well.
Which you can see in the latest commit

**Review by bvtujo — APPROVED** [review]
LGTM

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts [hunk 4]
```diff
@@ -196,6 +196,14 @@ export interface NetworkLoadBalancedTaskImageProps {
...
    */
   readonly dockerLabels?: { [key: string]: string };
+  /**
+   * The amount (in GiB) of ephemeral storage to be allocated to the task. The maximum supported value is 200GiB.
+   *
+   * NOTE: This parameter is only supported for tasks hosted on AWS Fargate using platform version 1.4.0 or later.
+   *
+   * @default - Undefined, in which case, the task will receive 20GiB ephemeral storage.
+   */
+  readonly ephemeralStorageGiB?: number;
```
Comment (corymhall) [inline packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:206]: Would it make sense to put this property on the `Fargate` interfaces rather than
the base interfaces? Then you wouldn't need to checks under the `ecs` services.

**Review by corymhall — DISMISSED** [review]
Just one question

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts (hunk 4 above)
Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:206]: Actually we can do that. But Where we closely observed the `taskImageOptions` were using in both fargate and ecs service patterns except in QueueProcessing. So I thought that its better to use the above property in the base interface rather than Fargate interfaces and pass the validation for ecs service patterns that when a customer is trying to use ephemeral storage by using ecs service pattern then it throws error saying "ephemeralStorageGiB is only supported for Fargate services."

When you see for  QueueProcessing where I put above property in the Fargate interfaces rather than
the base interfaces only. Because this was not using TaskImageOptions.

Comment (corymhall) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:206]: Yeah I think we should now have separate taskImageOptions for ecs vs fargate since the properties that can be input are now different. From a user experience perspective it's better to not be able to input invalid properties than to get an error saying that you can't use a particular property.

Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:206]: OK so do I need to remove that validation in ecs service patterns and add this property directly into fargate service patterns?

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts [hunk 5]
```diff
@@ -29,6 +29,15 @@ export interface ApplicationLoadBalancedFargateServiceProps extends ApplicationL
...
   readonly securityGroups?: ISecurityGroup[];
+
+  /**
+   * The amount (in GiB) of ephemeral storage to be allocated to the task. The maximum supported value is 200GiB.
+   *
+   * NOTE: This parameter is only supported for tasks hosted on AWS Fargate using platform version 1.4.0 or later.
+   *
+   * @default - Undefined, in which case, the task will receive 20GiB ephemeral storage.
+   */
+  readonly ephemeralStorageGiB?: number;
```
Comment (corymhall) [inline packages/aws-cdk-lib/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts:40]: Sorry for the confusion, I meant that you should be able to create a new
`ApplicationLoadBalancedFargateTaskImageOptions`. So you would have something
like

```ts
export interface FargateTaskImageOptions extends ApplicationLoadBalancedTaskImageOptions {
  readonly ephemeralStorageGiB?: number
    }

/**
 * The properties for the ApplicationLoadBalancedFargateService service.
 */
export interface ApplicationLoadBalancedFargateServiceProps extends ApplicationLoadBalancedServiceBaseProps, FargateServiceBaseProps {
  ...

  /**
   * The properties required to create a new task definition. TaskDefinition or TaskImageOptions must be specified, but not both.
   *
   * @default none
   */
  readonly taskImageOptions?: FargateTaskImageOptions;
}
```

**Review by corymhall — CHANGES_REQUESTED** [review]

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts (hunk 5 above)
Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts]: Ya Gotcha. I updated as per your suggestion.
Please check it in the latest commit.

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts [hunk 6]
```diff
@@ -35,7 +35,7 @@ export interface NetworkMultipleTargetGroupsServiceBaseProps {
    *
    * @default - none
    */
-  readonly taskImageOptions?: NetworkLoadBalancedTaskImageProps;
+  readonly taskImageOptions?: NetworkLoadBalancedFargateTaskImageProps;
```
Comment (corymhall) [inline packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts:38]: You shouldn't need to update this since the base class doesn't need the fargate inputs.

Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/base/network-multiple-target-groups-service-base.ts]: Ok so I need not update the taskImageOptions since the base class doesn't need the fargate inputs.
So I updated the code by add following interface in fargate-service-base.ts 
and we can import that interface to the fargate patterns and use ephemeral storage in it 
```ts 
export interface FargateTaskImageOptions {
  readonly ephemeralStorageGiB?: number;
}
```
Changes you can see in the latest commit

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts [hunk 7]
```diff
@@ -4,11 +4,11 @@ import { FeatureFlags } from '../../../core';
 import * as cxapi from '../../../cx-api';
 import { Construct } from 'constructs';
 import { ApplicationLoadBalancedServiceBase, ApplicationLoadBalancedServiceBaseProps } from '../base/application-load-balanced-service-base';
-import { FargateServiceBaseProps } from '../base/fargate-service-base';
+import { FargateServiceBaseProps, FargateTaskImageOptions } from '../base/fargate-service-base';
 /**
  * The properties for the ApplicationLoadBalancedFargateService service.
  */
-export interface ApplicationLoadBalancedFargateServiceProps extends ApplicationLoadBalancedServiceBaseProps, FargateServiceBaseProps {
+export interface ApplicationLoadBalancedFargateServiceProps extends ApplicationLoadBalancedServiceBaseProps, FargateServiceBaseProps, FargateTaskImageOptions {
```
Comment (corymhall) [inline packages/aws-cdk-lib/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts:11]:
```suggestion
export interface ApplicationLoadBalancedFargateServiceProps extends ApplicationLoadBalancedServiceBaseProps, FargateServiceBaseProps {
```

The `ApplicationLoadBalancedFargateServiceProps` should not extend
TaskImageOptions. You should add a `taskImageOptions` property to this
interface.

**Review by corymhall — CHANGES_REQUESTED** [review]

File: packages/aws-cdk-lib/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts (hunk 7 above)
Comment (homakk) [reply packages/aws-cdk-lib/aws-ecs-patterns/lib/fargate/application-load-balanced-fargate-service.ts:11]: It is not working actually when I try that it throwing the following error, where taskImageOptions was already existing in 
ApplicationLoadBalancedTaskImageOptions

image
image
